File: modules/background-location/android/src/main/java/app/koord/backgroundlocation/LocationDbHelper.kt
package app.koord.backgroundlocation

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

class LocationDbHelper(private val context: Context) {

    private val DATABASE_NAME = "workhourstracker.db"
    private val TABLE_NAME = "local_location_events"

    // Column names
    private val COLUMN_ID = "id"
    private val COLUMN_TIMESTAMP = "timestamp"
    private val COLUMN_COMPANY_ID = "company_id"
    private val COLUMN_TYPE = "type"
    private val COLUMN_ASSIGNMENT_ID = "assignment_id"
    private val COLUMN_WORKER_ID = "worker_id"
    private val COLUMN_LATITUDE = "latitude"
    private val COLUMN_LONGITUDE = "longitude"
    private val COLUMN_NOTES = "notes"
    private val COLUMN_SYNCED = "synced"
    private val TRANSITION_EVENT_TYPES = setOf("enter_geofence", "exit_geofence")

    private fun getDatabaseFile(): File {
        return context.getDatabasePath(DATABASE_NAME)
    }

    private fun openOrCreateAndPrepareDatabase(): SQLiteDatabase {
        val databaseFile = getDatabaseFile()
        val parentDir = databaseFile.parentFile
        if (parentDir != null && !parentDir.exists()) {
            val created = parentDir.mkdirs()
            if (created) {
                Log.d("LocationDbHelper", "Created database directory: ${parentDir.absolutePath}")
            } else {
                Log.e("LocationDbHelper", "Failed to create database directory: ${parentDir.absolutePath}")
            }
        }

        val db = SQLiteDatabase.openOrCreateDatabase(databaseFile.absolutePath, null)
        
        // Ensure table exists
        if (!tableExists(db, TABLE_NAME)) {
            db.execSQL(CREATE_LOCAL_LOCATION_EVENTS_TABLE_SQL)
            Log.d("LocationDbHelper", "Table $TABLE_NAME created.")
        } else if (!columnExists(db, TABLE_NAME, COLUMN_COMPANY_ID)) {
            rebuildLocationEventsTable(db, copyLegacyRows = false)
            Log.d("LocationDbHelper", "Rebuilt $TABLE_NAME without copying legacy rows missing $COLUMN_COMPANY_ID.")
        } else if (!columnIsNotNull(db, TABLE_NAME, COLUMN_COMPANY_ID)) {
            rebuildLocationEventsTable(db, copyLegacyRows = true)
            Log.d("LocationDbHelper", "Rebuilt $TABLE_NAME to require $COLUMN_COMPANY_ID.")
        }
        return db
    }

    fun insertLocationEvent(
        id: String,
        timestamp: Long, // Use long for Unix timestamp
        companyId: String,
        type: String,
        assignmentId: String,
        workerId: String,
        latitude: Double,
        longitude: Double,
        notes: String?,
        synced: Int // 0 for not synced
    ): Boolean {
        var db: SQLiteDatabase? = null
        try {
            db = openOrCreateAndPrepareDatabase()

            if (shouldSkipTransitionInsert(db, assignmentId, type)) {
                Log.d("LocationDbHelper", "Skipping duplicate transition $type for assignment $assignmentId")
                return false
            }
            
            val values = ContentValues().apply {
                put(COLUMN_ID, id)
                put(COLUMN_TIMESTAMP, SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).format(Date(timestamp))) // ISO 8601 string
                put(COLUMN_COMPANY_ID, companyId)
                put(COLUMN_TYPE, type)
                put(COLUMN_ASSIGNMENT_ID, assignmentId)
                put(COLUMN_WORKER_ID, workerId)
                put(COLUMN_LATITUDE, latitude)
                put(COLUMN_LONGITUDE, longitude)
                put(COLUMN_NOTES, notes)
                put(COLUMN_SYNCED, synced)
            }
            val result = db.insertWithOnConflict(TABLE_NAME, null, values, SQLiteDatabase.CONFLICT_IGNORE)
            if (result != -1L) {
                Log.d("LocationDbHelper", "Location event inserted successfully: $id")
                return true
            } else {
                Log.d("LocationDbHelper", "Location event already present or ignored: $id")
                return false
            }
        } catch (e: Exception) {
            Log.e("LocationDbHelper", "Error inserting location event: ${e.message}", e)
            return false
        } finally {
            db?.close()
        }
    }

    fun updateLocationEventSyncedStatus(id: String, syncedStatus: Int) {
        var db: SQLiteDatabase? = null
        try {
            db = openOrCreateAndPrepareDatabase()
            
            val values = ContentValues().apply {
                put(COLUMN_SYNCED, syncedStatus)
            }
            val rowsAffected = db.update(TABLE_NAME, values, "$COLUMN_ID = ?", arrayOf(id))
            if (rowsAffected > 0) {
                Log.d("LocationDbHelper", "Location event $id synced status updated to $syncedStatus")
            } else {
                Log.w("LocationDbHelper", "Failed to update synced status for location event $id. Event not found or no change.")
            }
        } catch (e: Exception) {
            Log.e("LocationDbHelper", "Error updating synced status for location event $id: ${e.message}", e)
        } finally {
            db?.close()
        }
    }

    private fun tableExists(db: SQLiteDatabase, tableName: String): Boolean {
        val cursor = db.rawQuery("SELECT DISTINCT tbl_name FROM sqlite_master WHERE tbl_name = '$tableName'", null)
        val exists = cursor.count > 0
        cursor.close()
        return exists
    }

    private fun columnExists(db: SQLiteDatabase, tableName: String, columnName: String): Boolean {
        val cursor = db.rawQuery("PRAGMA table_info($tableName)", null)
        val nameColumnIndex = cursor.getColumnIndex("name")
        while (cursor.moveToNext()) {
            if (nameColumnIndex >= 0 && cursor.getString(nameColumnIndex) == columnName) {
                cursor.close()
                return true
            }
        }
        cursor.close()
        return false
    }

    private fun columnIsNotNull(db: SQLiteDatabase, tableName: String, columnName: String): Boolean {
        val cursor = db.rawQuery("PRAGMA table_info($tableName)", null)
        val nameColumnIndex = cursor.getColumnIndex("name")
        val notNullColumnIndex = cursor.getColumnIndex("notnull")
        while (cursor.moveToNext()) {
            if (nameColumnIndex >= 0 && cursor.getString(nameColumnIndex) == columnName) {
                val isNotNull = notNullColumnIndex >= 0 && cursor.getInt(notNullColumnIndex) == 1
                cursor.close()
                return isNotNull
            }
        }
        cursor.close()
        return false
    }

    private fun rebuildLocationEventsTable(db: SQLiteDatabase, copyLegacyRows: Boolean) {
        db.beginTransaction()
        try {
            db.execSQL("ALTER TABLE $TABLE_NAME RENAME TO ${TABLE_NAME}_legacy")
            db.execSQL(CREATE_LOCAL_LOCATION_EVENTS_TABLE_SQL)
            if (copyLegacyRows) {
                db.execSQL(
                    """
                    INSERT INTO $TABLE_NAME (id, timestamp, company_id, type, assignment_id, worker_id, latitude, longitude, notes, synced)
                    SELECT id, timestamp, company_id, type, assignment_id, worker_id, latitude, longitude, notes, synced
                    FROM ${TABLE_NAME}_legacy
                    WHERE company_id IS NOT NULL
                    """.trimIndent()
                )
            }
            db.execSQL("DROP TABLE ${TABLE_NAME}_legacy")
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    private fun shouldSkipTransitionInsert(db: SQLiteDatabase, assignmentId: String, type: String): Boolean {
        if (!TRANSITION_EVENT_TYPES.contains(type)) {
            return false
        }

        val cursor = db.rawQuery(
            """
            SELECT $COLUMN_TYPE
            FROM $TABLE_NAME
            WHERE $COLUMN_ASSIGNMENT_ID = ?
              AND $COLUMN_TYPE IN (?, ?)
            ORDER BY $COLUMN_TIMESTAMP DESC
            LIMIT 1
            """.trimIndent(),
            arrayOf(assignmentId, "enter_geofence", "exit_geofence")
        )

        try {
            if (!cursor.moveToFirst()) {
                return false
            }
            val typeColumnIndex = cursor.getColumnIndex(COLUMN_TYPE)
            if (typeColumnIndex < 0) {
                return false
            }
            val lastTransitionType = cursor.getString(typeColumnIndex)
            return lastTransitionType == type
        } finally {
            cursor.close()
        }
    }

    // Re-using the SQL statement for local_location_events table creation from database.ts
    private val CREATE_LOCAL_LOCATION_EVENTS_TABLE_SQL = """
        CREATE TABLE IF NOT EXISTS local_location_events (
            id TEXT PRIMARY KEY NOT NULL,
            timestamp TEXT NOT NULL,
            company_id TEXT NOT NULL,
            type TEXT NOT NULL,
            assignment_id TEXT NOT NULL,
            worker_id TEXT NOT NULL,
            latitude REAL NOT NULL,
            longitude REAL NOT NULL,
            notes TEXT,
            synced INTEGER DEFAULT 0 NOT NULL
        );
    """.trimIndent()
}

File: modules/background-location/ios/LocationDbHelper.swift
import Foundation
import SQLite3

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

final class LocationDbHelper {
  static let shared = LocationDbHelper()

  private let tableName = "local_location_events"
  private let transitionEventTypes = Set(["enter_geofence", "exit_geofence"])

  private init() {}

  @discardableResult
  func insertLocationEvent(_ event: LocationEventRecord) -> Bool {
    guard let db = openDatabase() else { return false }
    defer { sqlite3_close(db) }

    guard prepareDatabase(db) else { return false }

    if shouldSkipTransitionInsert(db: db, assignmentId: event.assignmentId, type: event.type) {
      NSLog("LocationDbHelper: skipping duplicate transition \(event.type) for assignment \(event.assignmentId)")
      return false
    }

    let sql = """
      INSERT OR IGNORE INTO \(tableName)
      (id, timestamp, company_id, type, assignment_id, worker_id, latitude, longitude, notes, synced)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 0);
    """

    var statement: OpaquePointer?
    guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK else { return false }
    defer { sqlite3_finalize(statement) }

    bind(text: event.id, to: 1, in: statement)
    bind(text: isoString(from: event.createdAt), to: 2, in: statement)
    bind(text: event.companyId, to: 3, in: statement)
    bind(text: event.type, to: 4, in: statement)
    bind(text: event.assignmentId, to: 5, in: statement)
    bind(text: event.workerId, to: 6, in: statement)
    sqlite3_bind_double(statement, 7, event.latitude)
    sqlite3_bind_double(statement, 8, event.longitude)
    bind(text: event.notes, to: 9, in: statement)

    return sqlite3_step(statement) == SQLITE_DONE && sqlite3_changes(db) > 0
  }

  func markSynced(id: String) {
    guard let db = openDatabase() else { return }
    defer { sqlite3_close(db) }
    guard prepareDatabase(db) else { return }

    let sql = "UPDATE \(tableName) SET synced = 1 WHERE id = ?;"
    var statement: OpaquePointer?
    guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK else { return }
    defer { sqlite3_finalize(statement) }
    bind(text: id, to: 1, in: statement)
    sqlite3_step(statement)
  }

  private func openDatabase() -> OpaquePointer? {
    let fileManager = FileManager.default
    guard let documentsDirectory = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
      return nil
    }
    let sqliteDirectory = documentsDirectory.appendingPathComponent("SQLite", isDirectory: true)
    try? fileManager.createDirectory(at: sqliteDirectory, withIntermediateDirectories: true, attributes: nil)
    let dbURL = sqliteDirectory.appendingPathComponent("workhourstracker.db")

    var db: OpaquePointer?
    guard sqlite3_open(dbURL.path, &db) == SQLITE_OK else {
      if db != nil {
        sqlite3_close(db)
      }
      return nil
    }
    return db
  }

  private func prepareDatabase(_ db: OpaquePointer?) -> Bool {
    let createSQL = """
      CREATE TABLE IF NOT EXISTS \(tableName) (
        id TEXT PRIMARY KEY NOT NULL,
        timestamp TEXT NOT NULL,
        company_id TEXT NOT NULL,
        type TEXT NOT NULL,
        assignment_id TEXT NOT NULL,
        worker_id TEXT NOT NULL,
        latitude REAL NOT NULL,
        longitude REAL NOT NULL,
        notes TEXT,
        synced INTEGER DEFAULT 0 NOT NULL
      );
    """
    guard sqlite3_exec(db, createSQL, nil, nil, nil) == SQLITE_OK else { return false }
    return ensureCompanyIdIsRequired(db)
  }

  private func ensureCompanyIdIsRequired(_ db: OpaquePointer?) -> Bool {
    let pragmaSQL = "PRAGMA table_info(\(tableName));"
    var statement: OpaquePointer?
    guard sqlite3_prepare_v2(db, pragmaSQL, -1, &statement, nil) == SQLITE_OK else { return false }
    defer { sqlite3_finalize(statement) }

    var hasCompanyId = false
    var companyIdNotNull = false

    while sqlite3_step(statement) == SQLITE_ROW {
      if let nameC = sqlite3_column_text(statement, 1) {
        let name = String(cString: nameC)
        if name == "company_id" {
          hasCompanyId = true
          companyIdNotNull = sqlite3_column_int(statement, 3) == 1
        }
      }
    }

    guard !hasCompanyId || !companyIdNotNull else { return true }

    if !hasCompanyId {
      let rebuildSQL = """
        BEGIN TRANSACTION;
        ALTER TABLE \(tableName) RENAME TO \(tableName)_legacy;
        \(createTableSQL())
        DROP TABLE \(tableName)_legacy;
        COMMIT;
      """
      return sqlite3_exec(db, rebuildSQL, nil, nil, nil) == SQLITE_OK
    }

    let rebuildSQL = """
      BEGIN TRANSACTION;
      ALTER TABLE \(tableName) RENAME TO \(tableName)_legacy;
      \(createTableSQL())
      INSERT INTO \(tableName) (id, timestamp, company_id, type, assignment_id, worker_id, latitude, longitude, notes, synced)
      SELECT id, timestamp, company_id, type, assignment_id, worker_id, latitude, longitude, notes, synced
      FROM \(tableName)_legacy
      WHERE company_id IS NOT NULL;
      DROP TABLE \(tableName)_legacy;
      COMMIT;
    """
    return sqlite3_exec(db, rebuildSQL, nil, nil, nil) == SQLITE_OK
  }

  private func createTableSQL() -> String {
    """
    CREATE TABLE IF NOT EXISTS \(tableName) (
      id TEXT PRIMARY KEY NOT NULL,
      timestamp TEXT NOT NULL,
      company_id TEXT NOT NULL,
      type TEXT NOT NULL,
      assignment_id TEXT NOT NULL,
      worker_id TEXT NOT NULL,
      latitude REAL NOT NULL,
      longitude REAL NOT NULL,
      notes TEXT,
      synced INTEGER DEFAULT 0 NOT NULL
    );
    """
  }

  private func shouldSkipTransitionInsert(db: OpaquePointer?, assignmentId: String, type: String) -> Bool {
    guard transitionEventTypes.contains(type) else { return false }

    let sql = """
      SELECT type
      FROM \(tableName)
      WHERE assignment_id = ?
        AND type IN ('enter_geofence', 'exit_geofence')
      ORDER BY timestamp DESC
      LIMIT 1;
    """
    var statement: OpaquePointer?
    guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK else { return false }
    defer { sqlite3_finalize(statement) }
    bind(text: assignmentId, to: 1, in: statement)

    guard sqlite3_step(statement) == SQLITE_ROW,
          let typeC = sqlite3_column_text(statement, 0) else {
      return false
    }
    let lastType = String(cString: typeC)
    return lastType == type
  }

  private func bind(text: String?, to index: Int32, in statement: OpaquePointer?) {
    guard let text else {
      sqlite3_bind_null(statement, index)
      return
    }
    sqlite3_bind_text(statement, index, text, -1, SQLITE_TRANSIENT)
  }

  private func isoString(from date: Date) -> String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    return formatter.string(from: date)
  }
}
